pkg/capture/capture_manager.go

```diff
@@ -1,436 +1,355 @@
-/////////////////////////////////////////////////////////////////////////////////
-//
-// capture_manager.go
-//
-// Written by Lorenz Breidenbach lob@open.ch,
-//            Lennart Elsen lel@open.ch, December 2015
-// Copyright (c) 2015 Open Systems AG, Switzerland
-// All Rights Reserved.
-//
-/////////////////////////////////////////////////////////////////////////////////
-
 package capture
 
 import (
 	"context"
 	"fmt"
-	"sort"
 	"sync"
 	"time"
 
 	"github.com/els0r/goProbe/cmd/goProbe/config"
 	"github.com/els0r/goProbe/pkg/capture/capturetypes"
+	"github.com/els0r/goProbe/pkg/goDB"
+	"github.com/els0r/goProbe/pkg/goDB/encoder/encoders"
+	"github.com/els0r/goProbe/pkg/goprobe/writeout"
 	"github.com/els0r/goProbe/pkg/logging"
-	"github.com/els0r/goProbe/pkg/types/hashmap"
-)
-
-const (
-	// MaxIfaces is the maximum number of interfaces we can monitor
-	MaxIfaces = 1024
-
-	// WriteoutsChanDepth sets the maximum amount of writeouts that can be queued
-	WriteoutsChanDepth = 100
+	"golang.org/x/exp/slog"
 )
 
-// TaggedAggFlowMap represents an aggregated
-// flow map tagged with Stats and an
-// an interface name.
-//
-// Used by Manager to return the results of
-// RotateAll() and Update().
-type TaggedAggFlowMap struct {
-	Map   *hashmap.AggFlowMap
-	Stats capturetypes.PacketStats `json:"stats,omitempty"`
-	Iface string                   `json:"iface"`
-}
+const allowedWriteoutDurationFraction = 0.1
 
 // Manager manages a set of Capture instances.
 // Each interface can be associated with up to one Capture.
 type Manager struct {
-	sync.Mutex
-	captures     map[string]*ManagedCapture
-	ctx          context.Context
-	sourceInitFn sourceInitFn
-}
-
-type ManagedCapture struct {
-	capture *Capture
-	cancel  context.CancelFunc
-}
+	sync.RWMutex
```

Comment:
This mutex is now solely responsible for locking the external methods (i.e. `Update()` / `Close()` (which under the hood is actually using `Update()` and access to the last rotation). Locking of the individual captures within the RunGroups is done by the new methods introduced for the `Captures` type.